Document log based alerting (#1521)
giantswarm/prometheus-rules: README.md

The alerting rules are located in `helm/prometheus-rules/templates/<area>/<team>/alerting-rules` in the specific area/team to which they belong.

### Alert structure

Giant Swarm supports two different types of alert rules:

> #### 1. Metrics-based alerts

Reviewer: ```suggestion
#### Metrics-based alerts
```

These are standard Prometheus alerts based on PromQL queries against metrics data. They are stored in files ending with `.rules.yaml`.

Example:

> ```yaml

Reviewer: ```suggestion
```yaml
# management-cluster-app-failed-atlas.rules.yaml
```

apiVersion: monitoring.coreos.com/v1
kind: PrometheusRule
metadata:
  name: management-cluster-app-failed-atlas
  namespace: monitoring
  labels:
    area: platform
    team: atlas
spec:
  groups:
  - name: app
    rules:
    - alert: ManagementClusterAppFailedAtlas
      annotations:
        summary: Management cluster app not deployed correctly
        description: '{{`Management Cluster App {{ $labels.name }}, version {{ $labels.version }} is {{if $labels.status }} in {{ $labels.status }} state. {{else}} not installed. {{end}}`}}'

>         __dashboardUid__: UniqueID of the dashboard

Reviewer: unique-id-of-the-dashboard

Author: ```suggestion
        __dashboardUid__: unique-id-of-the-dashboard
```

>         __panelId__: id of the panel in the dashboard

Reviewer: ```suggestion
        __panelId__: 42 # id of the panel in the dashboard
```

        dashboardQueryParams: "orgid=1"
        # dashboardExternalUrl: https://link-to-my-dashboard
        runbook_url: https://intranet.giantswarm.io/docs/support-and-ops/ops-recipes/app-failed/
      expr: app_operator_app_info{status!~"(?i:(deployed|cordoned))", catalog=~"control-plane-.*",team="atlas"}
      for: 30m
      labels:
        area: platform
        cancel_if_cluster_status_creating: "true"
        cancel_if_cluster_status_deleting: "true"
        cancel_if_cluster_status_updating: "true"
        cancel_if_outside_working_hours: "true"
        severity: page
        sig: none
        team: atlas
```

> #### 2. Logs-based alerts

Reviewer: ```suggestion
#### Logs-based alerts
```

These alerts are generated from log data using LogQL queries processed by Loki. To create a log-based alert:

> - Add the label `application.giantswarm.io/prometheus-rule-kind: loki` to the alert

Reviewer: Also why is there both a label and a naming convention for LogQL based alerts ?

Author: I think it will be easier to see them in a folder and also we need some differentiator when we add the testing tooling

Reviewer: Ok so naming is for the repository organization and the label for deployment purposes.

Reviewer 2: But maybe helm could automatically add the label for files with this extension?

Author: You're right, I edited the _helpers.tpl to add the alert label if the files ends with `logs.yml` or `logs.yaml`

        runbook_url: https://intranet.giantswarm.io/docs/support-and-ops/ops-recipes/log-errors/
      expr: sum(rate({app="my-app"} |= "error" [5m])) > 100
      for: 10m
      labels:
        severity: page
        team: myteam
        area: platform
```

Log-based alerts are processed differently in the observability platform but appear alongside metrics-based alerts in alerting interfaces.

### Best practices

We follow standardized practices for organizing our alerts using PrometheusRule custom resources.

#### Mandatory annotations
- `description`: Detailed explanation of what happened and what the alert is monitoring
- `runbook_url`: Link to a runbook page with incident management instructions

#### Recommended annotations
- `summary`: Brief overview of what the alert detected
- `__dashboardUid__`: Unique identifier of the relevant dashboard
- `__panelId__`: Specific panel ID within the referenced dashboard
- `dashboardQueryParams`: Additional URL parameters (must include `orgId=1` for Shared Org dashboards or `orgId=2` for others)
- `dashboardExternalUrl`: Optional link to an external Grafana instance (like Grafana Cloud)

##### Dashboard URL construction

> Alert manager generates dashboard URLs for Opsgenie and Slack alerts using these rules:

Reviewer: ```suggestion
Alertmanager generates dashboard URLs for Opsgenie and Slack alerts using these rules:
```

1. With only `__dashboardUid__`: `https://grafana.domain/__dashboardUid__`
2. With both `__dashboardUid__` and `dashboardQueryParams`: `https://grafana.domain/__dashboardUid__?dashboardQueryParams`
3. If `dashboardExternalUrl` is set: Uses the exact URL provided

#### Mandatory labels
- `area`: Functional area (e.g., platform, apps)
- `team`: Responsible team identifier
- `severity`: Alert severity level (page, notify)

#### Optional labels
- `cancel_if_*`: Labels used for alert inhibitions

> - `all_pipelines: "true"`: Ensures the alert is sent to Opsgenie regardless of installation stability status

Reviewer: ```suggestion
- `all_pipelines: "true"`: Ensures the alert is sent to Opsgenie regardless of installation's pipeline
```

        description: '{{`Management Cluster App {{ $labels.name }}, version {{ $labels.version }} is {{if $labels.status }} in {{ $labels.status }} state. {{else}} not installed. {{end}}`}}'
        __dashboardUid__: UniqueID of the dashboard
        __panelId__: id of the panel in the dashboard
        dashboardQueryParams: "orgid=1"
        runbook_url: https://intranet.giantswarm.io/docs/support-and-ops/ops-recipes/app-failed/
      expr: app_operator_app_info{status!~"(?i:(deployed|cordoned))", catalog=~"control-plane-.*",team="atlas"}
      for: 30m
      labels:
        area: platform
        cancel_if_cluster_status_creating: "true"
        cancel_if_cluster_status_deleting: "true"
        cancel_if_cluster_status_updating: "true"
        cancel_if_outside_working_hours: "true"
        severity: page
        sig: none
        team: atlas
```

#### 2. Logs-based alerts

These alerts are generated from log data using LogQL queries processed by Loki. To create a log-based alert:

- Add the label `application.giantswarm.io/prometheus-rule-kind: loki` to the alert
- Ensure the `expr` field contains a valid LogQL query

> - Name the file with a `.logs.yaml` extension
> Example:

Reviewer 3: ```suggestion
- Name the file with a `.logs.yaml` extension

Example:
```

> ```yaml
> apiVersion: monitoring.coreos.com/v1

Reviewer 3: If the filename matters, consider including that in the example here
```suggestion
```yaml
# log-based-alerts-example.logs.yaml
apiVersion: monitoring.coreos.com/v1
```

Author: I don't think it matters tbh but I would prefer that we have some conventions if we add more kinds of alerts and this will most likely help with testing

Reviewer 3: I ask only because it says
> Name the file with a .logs.yaml extension

which makes me think that is required

### Best practices

We follow standardized practices for organizing our alerts using PrometheusRule custom resources.

#### Mandatory annotations
- `description`: Detailed explanation of what happened and what the alert is monitoring
- `runbook_url`: Link to a runbook page with incident management instructions

#### Recommended annotations
- `summary`: Brief overview of what the alert detected
- `__dashboardUid__`: Unique identifier of the relevant dashboard
- `__panelId__`: Specific panel ID within the referenced dashboard
- `dashboardQueryParams`: Additional URL parameters (must include `orgId=1` for Shared Org dashboards or `orgId=2` for others)
- `dashboardExternalUrl`: Optional link to an external Grafana instance (like Grafana Cloud)

##### Dashboard URL construction

Alert manager generates dashboard URLs for Opsgenie and Slack alerts using these rules:

1. With only `__dashboardUid__`: `https://grafana.domain/__dashboardUid__`
2. With both `__dashboardUid__` and `dashboardQueryParams`: `https://grafana.domain/__dashboardUid__?dashboardQueryParams`

> 3. If `dashboardExternalUrl` is set: Uses the exact URL provided

Reviewer 3: Maybe include an example in the sample yaml?

Author: For external url? 🤔 good point

Reviewer 3: Ah, I was looking only at the log alert example and didn't see the dashboard in the metrics example. That works too

1. With only `__dashboardUid__`: `https://grafana.domain/__dashboardUid__`
2. With both `__dashboardUid__` and `dashboardQueryParams`: `https://grafana.domain/__dashboardUid__?dashboardQueryParams`
3. If `dashboardExternalUrl` is set: Uses the exact URL provided

#### Mandatory labels
- `area`: Functional area (e.g., platform, apps)
- `team`: Responsible team identifier
- `severity`: Alert severity level (page, notify)

> - `provider`: Cloud provider

Author: ```suggestion
```

> - `provider`: Cloud provider

Reviewer 3: Are these added automatically by the o11yplat?

Author: They are today yes (by the scraping agent) but we know provider is not always accurate

Author: I would but it;s not added by our logging pipeline :D

Author: I think I will just remove them actually

#### Mandatory annotations
- `description`: Detailed explanation of what happened and what the alert is monitoring
- `runbook_url`: Link to a runbook page with incident management instructions

#### Recommended annotations
- `summary`: Brief overview of what the alert detected
- `__dashboardUid__`: Unique identifier of the relevant dashboard
- `__panelId__`: Specific panel ID within the referenced dashboard
- `dashboardQueryParams`: Additional URL parameters (must include `orgId=1` for Shared Org dashboards or `orgId=2` for others)
- `dashboardExternalUrl`: Optional link to an external Grafana instance (like Grafana Cloud)

##### Dashboard URL construction

Alert manager generates dashboard URLs for Opsgenie and Slack alerts using these rules:

1. With only `__dashboardUid__`: `https://grafana.domain/__dashboardUid__`
2. With both `__dashboardUid__` and `dashboardQueryParams`: `https://grafana.domain/__dashboardUid__?dashboardQueryParams`
3. If `dashboardExternalUrl` is set: Uses the exact URL provided

#### Mandatory labels
- `area`: Functional area (e.g., platform, apps)
- `team`: Responsible team identifier
- `severity`: Alert severity level (page, notify)

> - `pipeline`: Deployment pipeline

Author: ```suggestion
```

The alerting rules are located in `helm/prometheus-rules/templates/<area>/<team>/alerting-rules` in the specific area/team to which they belong.

### Alert structure

Giant Swarm supports two different types of alert rules:

#### 1. Metrics-based alerts

These are standard Prometheus alerts based on PromQL queries against metrics data. They are stored in files ending with `.rules.yaml`.

Example:

```yaml
apiVersion: monitoring.coreos.com/v1
kind: PrometheusRule
metadata:
  name: management-cluster-app-failed-atlas
  namespace: monitoring
  labels:
    area: platform
    team: atlas
spec:
  groups:
  - name: app
    rules:
    - alert: ManagementClusterAppFailedAtlas
      annotations:
        summary: Management cluster app not deployed correctly
        description: '{{`Management Cluster App {{ $labels.name }}, version {{ $labels.version }} is {{if $labels.status }} in {{ $labels.status }} state. {{else}} not installed. {{end}}`}}'
        __dashboardUid__: UniqueID of the dashboard
        __panelId__: id of the panel in the dashboard

>         dashboardQueryParams: "orgid=1"

Author: ```suggestion
        dashboardQueryParams: "orgid=1"
        # dashboardExternalUrl: https://link-to-my-dashboard
```

Author: @stone-z I added it here

#### Mandatory annotations
- `description`: Detailed explanation of what happened and what the alert is monitoring
- `runbook_url`: Link to a runbook page with incident management instructions

#### Recommended annotations
- `summary`: Brief overview of what the alert detected
- `__dashboardUid__`: Unique identifier of the relevant dashboard
- `__panelId__`: Specific panel ID within the referenced dashboard
- `dashboardQueryParams`: Additional URL parameters (must include `orgId=1` for Shared Org dashboards or `orgId=2` for others)
- `dashboardExternalUrl`: Optional link to an external Grafana instance (like Grafana Cloud)

##### Dashboard URL construction

Alert manager generates dashboard URLs for Opsgenie and Slack alerts using these rules:

1. With only `__dashboardUid__`: `https://grafana.domain/__dashboardUid__`
2. With both `__dashboardUid__` and `dashboardQueryParams`: `https://grafana.domain/__dashboardUid__?dashboardQueryParams`
3. If `dashboardExternalUrl` is set: Uses the exact URL provided

#### Mandatory labels
- `area`: Functional area (e.g., platform, apps)
- `team`: Responsible team identifier
- `severity`: Alert severity level (page, notify)

> - `cluster_id`: Associated cluster ID

Author: ```suggestion
```

- `description`: Detailed explanation of what happened and what the alert is monitoring
- `runbook_url`: Link to a runbook page with incident management instructions

#### Recommended annotations
- `summary`: Brief overview of what the alert detected
- `__dashboardUid__`: Unique identifier of the relevant dashboard
- `__panelId__`: Specific panel ID within the referenced dashboard
- `dashboardQueryParams`: Additional URL parameters (must include `orgId=1` for Shared Org dashboards or `orgId=2` for others)
- `dashboardExternalUrl`: Optional link to an external Grafana instance (like Grafana Cloud)

##### Dashboard URL construction

Alert manager generates dashboard URLs for Opsgenie and Slack alerts using these rules:

1. With only `__dashboardUid__`: `https://grafana.domain/__dashboardUid__`
2. With both `__dashboardUid__` and `dashboardQueryParams`: `https://grafana.domain/__dashboardUid__?dashboardQueryParams`
3. If `dashboardExternalUrl` is set: Uses the exact URL provided

#### Mandatory labels
- `area`: Functional area (e.g., platform, apps)
- `team`: Responsible team identifier
- `severity`: Alert severity level (page, notify)
- `cluster_id`: Associated cluster ID

> - `installation`: Installation identifier

Author: ```suggestion
```
